File: cpu/cpp/src/quicksort.cpp
/**
 * QuickSort Benchmark — C++
 * ==========================
 * Benchmarks std::sort (introsort) and manual quicksort on integer arrays.
 * Input sizes: 1M, 10M, 100M elements.
 * Compiled with -O3 for fair comparison against Rust's slice::sort_unstable.
 */

#include <algorithm>
#include <benchmark/benchmark.h>
#include <cstdint>
#include <numeric>
#include <random>
#include <vector>

#include "benchmark_utils.h"

// ── Manual QuickSort Implementation ─────────────────────────────────────────

namespace {

template <typename Iter>
Iter partition(Iter first, Iter last) {
    auto pivot = std::prev(last);
    auto i = first;
    for (auto j = first; j != pivot; ++j) {
        if (*j <= *pivot) {
            std::iter_swap(i, j);
            ++i;
        }
    }
    std::iter_swap(i, pivot);
    return i;
}

template <typename Iter>
void quicksort_impl(Iter first, Iter last) {
    if (std::distance(first, last) <= 1) return;
    
    // Use median-of-three for pivot selection
    auto mid = first + std::distance(first, last) / 2;
    auto pivot_candidate = std::prev(last);
    
    if (*mid < *first) std::iter_swap(mid, first);
    if (*pivot_candidate < *first) std::iter_swap(pivot_candidate, first);
    if (*mid < *pivot_candidate) std::iter_swap(mid, pivot_candidate);
    
    auto pivot = partition(first, last);
    quicksort_impl(first, pivot);
    quicksort_impl(std::next(pivot), last);
}

} // anonymous namespace

// ── Benchmark: std::sort ────────────────────────────────────────────────────

static void BM_StdSort(benchmark::State& state) {
    const size_t n = state.range(0);
    auto original = bench_utils::generate_random_array<int32_t>(n);
    
    for (auto _ : state) {
        // Copy data each iteration (sorting is destructive)
        auto data = original;
        std::sort(data.begin(), data.end());
        benchmark::DoNotOptimize(data.data());
        benchmark::ClobberMemory();
    }
    
    state.SetItemsProcessed(static_cast<int64_t>(state.iterations()) * n);
    state.SetLabel("std::sort");
}

// ── Benchmark: Manual QuickSort ─────────────────────────────────────────────

static void BM_ManualQuickSort(benchmark::State& state) {
    const size_t n = state.range(0);
    auto original = bench_utils::generate_random_array<int32_t>(n);
    
    for (auto _ : state) {
        auto data = original;
        quicksort_impl(data.begin(), data.end());
        benchmark::DoNotOptimize(data.data());
        benchmark::ClobberMemory();
    }
    
    state.SetItemsProcessed(static_cast<int64_t>(state.iterations()) * n);
    state.SetLabel("manual_quicksort");
}

// ── Register Benchmarks ─────────────────────────────────────────────────────

BENCHMARK(BM_StdSort)
    ->Unit(benchmark::kMillisecond)
    ->Arg(1'000'000)
    ->Arg(10'000'000)
    ->Arg(100'000'000)
    ->MinTime(bench_utils::cpu_min_time_seconds())
    ->Repetitions(bench_utils::cpu_repetitions(30))
    ->ReportAggregatesOnly(true)
    ->DisplayAggregatesOnly(true);

BENCHMARK(BM_ManualQuickSort)
    ->Unit(benchmark::kMillisecond)
    ->Arg(1'000'000)
    ->Arg(10'000'000)
    ->Arg(100'000'000)
    ->MinTime(bench_utils::cpu_min_time_seconds())
    ->Repetitions(bench_utils::cpu_repetitions(30))
    ->ReportAggregatesOnly(true)
    ->DisplayAggregatesOnly(true);

File: cpu/rust/src/quicksort.rs
//! QuickSort Benchmark — Rust
//! ===========================
//! Uses `slice::sort_unstable` (pattern-defeating quicksort).
//! This is Rust's recommended sorting for types without expensive equality.

/// Sort using Rust's built-in sort_unstable (pdqsort).
/// This is the idiomatic, safe Rust approach.
pub fn std_sort_unstable(data: &mut [i32]) {
    data.sort_unstable();
}

/// Manual quicksort with median-of-three pivot selection.
/// Matches the C++ manual implementation for kernel parity.
pub fn manual_quicksort(data: &mut [i32]) {
    if data.len() <= 1 {
        return;
    }
    quicksort_inner(data);
}

fn quicksort_inner(data: &mut [i32]) {
    if data.len() <= 16 {
        // Insertion sort for small slices
        insertion_sort(data);
        return;
    }

    // Median-of-three pivot
    let len = data.len();
    let mid = len / 2;
    let last = len - 1;

    if data[mid] < data[0] {
        data.swap(0, mid);
    }
    if data[last] < data[0] {
        data.swap(0, last);
    }
    if data[mid] < data[last] {
        data.swap(mid, last);
    }

    // Partition using the last element as pivot
    let pivot = data[last];
    let mut i = 0;
    for j in 0..last {
        if data[j] <= pivot {
            data.swap(i, j);
            i += 1;
        }
    }
    data.swap(i, last);

    // Recurse
    let (left, right) = data.split_at_mut(i);
    quicksort_inner(left);
    if right.len() > 1 {
        quicksort_inner(&mut right[1..]);
    }
}

fn insertion_sort(data: &mut [i32]) {
    for i in 1..data.len() {
        let key = data[i];
        let mut j = i;
        while j > 0 && data[j - 1] > key {
            data[j] = data[j - 1];
            j -= 1;
        }
        data[j] = key;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::generate_random_i32;

    #[test]
    fn test_std_sort() {
        let mut data = generate_random_i32(10_000, 42);
        std_sort_unstable(&mut data);
        assert!(data.windows(2).all(|w| w[0] <= w[1]));
    }

    #[test]
    fn test_manual_quicksort() {
        let mut data = generate_random_i32(10_000, 42);
        manual_quicksort(&mut data);
        assert!(data.windows(2).all(|w| w[0] <= w[1]));
    }
}
